shared/teams/invite-by-email/container.native.js

```diff
@@ -0,0 +1,170 @@
+// @flow
+import * as Creators from '../../actions/teams/creators'
+import {Set, Map} from 'immutable'
+import InviteByEmailMobile from '.'
+import {HeaderHoc} from '../../common-adapters'
+import {navigateAppend} from '../../actions/route-tree'
+import * as Contacts from 'react-native-contacts'
+import {
+  connect,
+  compose,
+  withHandlers,
+  withPropsOnChange,
+  withState,
+  lifecycle,
+  type TypedState,
+} from '../../util/container'
+import {type OwnProps} from './container'
+import {isAndroid} from '../../constants/platform'
+
+const mapStateToProps = (state: TypedState, {routeProps}: OwnProps) => ({
+  name: routeProps.get('teamname'),
+  _pendingInvites: state.entities.getIn(
+    ['teams', 'teamNameToInvites', routeProps.get('teamname') || ''],
+    Set()
+  ),
+  loadingInvites: state.entities.getIn(
+    ['teams', 'teamNameToLoadingInvites', routeProps.get('teamname') || ''],
+    Map()
+  ),
+})
+
+const mapDispatchToProps = (dispatch: Dispatch, {navigateUp, routeProps}) => ({
+  onClose: () => dispatch(navigateUp()),
+  onInviteEmail: ({invitee, role}) => {
+    dispatch(Creators.inviteToTeamByEmail(routeProps.get('teamname'), role, invitee))
+    dispatch(Creators.getTeams())
+  },
+  onInvitePhone: ({invitee, role}) => {
+    dispatch(Creators.inviteToTeamByPhone(routeProps.get('teamname'), role, invitee))
+    dispatch(Creators.getTeams())
+  },
+  onUninvite: (invitee: string) => {
+    dispatch(Creators.removeMember(invitee, routeProps.get('teamname'), ''))
+  },
+
+  onOpenRolePicker: (role: string, onComplete: string => void) => {
+    dispatch(
+      navigateAppend([
+        {
+          props: {
+            allowOwner: false,
+            allowAdmin: false,
+            onComplete,
+            selectedRole: role,
+          },
+          selected: 'controlledRolePicker',
+        },
+      ])
+    )
+  },
+})
+
+export default compose(
+  connect(mapStateToProps, mapDispatchToProps),
+  compose(
+    withState('role', 'onRoleChange', 'writer'),
+    withPropsOnChange(['onExitSearch'], props => ({
+      onBack: () => props.onClose(),
+      title: 'Invite contacts',
+      headerStyle: {borderBottomWidth: 0},
+    })),
+    withState('contacts', '_setContacts', []),
+    withState('hasPermission', '_setHasPermission', true),
+    lifecycle({
+      componentWillMount() {
+        // TODO test this permission flow on a real build
+        if (isAndroid) {
+          Contacts.requestPermission((_, granted) => {
```

Comment:
maybe we should pull these out into little helpers so its not all inline in these big blocks